src/spartan/mod.rs

```diff
@@ -38,29 +38,43 @@ pub struct PolyEvalWitness<G: Group> {
 }
 
 impl<G: Group> PolyEvalWitness<G> {
-  fn pad(W: &[PolyEvalWitness<G>]) -> Vec<PolyEvalWitness<G>> {
-    // determine the maximum size
-    if let Some(n) = W.iter().map(|w| w.p.len()).max() {
-      W.iter()
-        .map(|w| {
-          let mut p = vec![G::Scalar::ZERO; n];
-          p[..w.p.len()].copy_from_slice(&w.p);
-          PolyEvalWitness { p }
-        })
-        .collect()
-    } else {
-      Vec::new()
-    }
-  }
+  /// Given [Pᵢ] and [sᵢ], compute P = ∑ᵢ sᵢ⋅Pᵢ
+  ///
+  /// # Details
+  ///
+  /// We allow the input polynomials to have different sizes, and interpret smaller ones as
+  /// being padded with 0 to the maximum size of all polynomials.
+  fn batch_diff_size(W: Vec<PolyEvalWitness<G>>, s: G::Scalar) -> PolyEvalWitness<G> {
```

Comment:
You can probably get away with the following signature:
```rust
  fn batch_diff_size(W: &mut [PolyEvalWitness<G>], s: G::Scalar) -> PolyEvalWitness<G> 
```

Reply:
To be honest, I think the signature should depend on whether we need to use the input polynomials as "scratch" buffer. My intuition is yes, since multiplying the `p_i`s independently by `s_i` first and then summing them has nicer memory access patterns over multiple threads. With the current implementation which overwrites the input polynomials with temporary computation, they are no longer be usable after the function is done. Therefore, I feel like it makes sense to consume them. What are the advantages of taking a `&mut` input?

Reply:
> What are the advantages of taking a &mut input?

Passing the inputs to the computation when someone else owns the allocation you're about to make, admittedly, unusable. It's just a composability / capability thing.